atomix/cluster/src/test/java/io/atomix/raft/SnapshotReplicationListenerTest.java

```diff
@@ -100,4 +100,18 @@ public void shouldNotCallListenerOnRegister() {
     verify(snapshotReplicationListener, times(0))
         .onSnapshotReplicationCompleted(follower.getTerm());
   }
+
+  @Test
+  public void shouldNotCallListenerOnRegisterIfLeader() {
```

Comment:
:thought_balloon: Can we test this with property based testing, using the infrastructure from `RandomizedRaftTest`? It might help us to find other race conditions, if there any. I don't have a concrete idea how it would look like. But might we worth considering.